geps/gep-2257.md

```diff
@@ -0,0 +1,100 @@
+Durations are a subset of what `time.ParseDuration` supports:
+
+- A GEP-2257 Duration MUST be one to four _components_, each of which consists
+  of an integer _value_ followed immediately by a _unit_. For example, `1h`,
+  `5m`, `1h5m`, and `1h30m30s500ms` are all valid GEP-2257 Durations.
+
+- For each component, the value MUST be one to five decimal digits. Floating
+  point is not allowed. Leading zeroes do not mean octal; the value MUST
+  always be interpreted as a decimal integer. For example, `1h`, `60m`, `01h`,
+  and `00060m` are all equivalent GEP-2257 Durations.
+
+- For each component, the unit MUST be one of `h` (hour), `m` (minute), `s`
+  (second), or `ms` (millisecond). No units larger than hours or smaller than
+  milliseconds are supported.
+
+- The total duration expressed by a GEP-2257 Duration string is the sum of
+  each of its components. For example, `1h30m` would be a 90-minute duration,
+  and `1s500ms` would be a 1.5-second duration.
+
+- There is no requirement that all units must be used. A GEP-2257 Duration of
+  `1h500ms` is supported (although probably not terribly useful).
+
+- Units MAY be repeated, although users SHOULD NOT rely on this support since
+  this GEP is Experimental, and future revisions may remove support for
+  repeated units. If units are repeated, the total duration remains the sum of
+  all components: a GEP-2257 duration of `1h2h20m10m` is a duration of 3 hours
+  30 minutes.
+
+- Since the value and the unit are both required within a component, `0` is
+  not a valid GEP-2257 duration string (though `0s` is). Likewise the empty
+  string is not a valid GEP-2257 duration.
+
+A GEP-2257 Duration parser can be easily implemented by doing a regex-match
+check before calling a parser equivalent to Go's `time.ParseDuration`. Such
+parsers are readily available in (at least) Go, Rust's `kube-rs` crate, and
+Python's `durationpy` package. We expect that these three languages cover the
+vast majority of the Kubernetes ecosystem.
+
+## Alternatives
+
+We considered three main alternatives:
+
+- Raw Golang `time.ParseDuration` format. This is very widely used in the Go
+  ecosystem -- however, it is a very open-ended specification and, in
+  particular, its support for floating-point values and negative durations
+  makes it difficult to validate.
+
+- Golang `strfmt.ParseDuration` as used in the APIServer's OpenAPI validation
+  code. It turns out that `strfmt.ParseDuration` is a superset of
+  `time.ParseDuration`, so all the problems in validation are still present.
+  Additionally, `strfmt.ParseDuration` supports day and week units, requiring
+  discussion of leap seconds.
+
+- ISO8601/RFC3339 durations. These are considerably less user-friendly than
+  our proposal: `PT0.5S` is simply not as immediately clear as "500ms".
+
+There is (a lot) more discussion in [PR 2155].
+
+[PR 2155]:https://github.com/kubernetes-sigs/gateway-api/pull/2155
```

Comment:
Very nice comparison with the other things we've considered, this is great for posterity for future readers.